### ISSUE TITLE
[feature request] In-place updates support for init containers in CloneSet

### ISSUE THREAD

**Issue by wgitg** [openkruise/kruise#560]
**What would you like to be added**:

We'd like to have in-place updates for init containers in CloneSet

**Why is this needed**:
In-place updates is an important requirement for us to migrate existing workload to Kubernetes (stabilities, maintainabilities, etc). We are levering CloneSet in-place update in the app containers. Currently we are also leveraging init containers to do some configurations that are needed before starting the applications. Updating image or anything else causes the pod to restart, which cancels the benefits of in-places updates support in the app container using CloneSet. So it'd be great if we can have in-place update support from init containers in CloneSet. Is it possible?

**Comment by wangyanghack** [comment]
@wgitg hi,I don't understand how it is necessary to have in-place update for init containers.I think main container is ready and running when you want to in-place update for init containers.Main container is still running after in-place update for init containers is completed.So init container after in-place update cannot execute again.Is it necessary to have in-place update for init containers?

**Comment by wgitg** [comment]
As mentioned in the Why section, but let us list out the use case in-place update in init container

Init containers execute before any main containers, and init containers can run as root to set up a lot of stuff that app container can not do. So they are used to set up anything needed before the app containers start. This can include getting applications certs, set up file structure, set up the file ownership/permission, etc.

So when the main containers are updated (e.g. version changed in `image` field) in-place, if init container does not support in-place, the pod will be recreated. For example, 

`pre-app-container-setup` is setting up the certs, files ownership before the app container start. Say first we deployed version 1.0 of `main-app` with the init container `pre-app-container-setup`, 

```
initContainers:
   - name: pre-app-container-setup
     image: pre-app-container-setup:100.0
     command: ["utils-bin"]
     args: [
        "--main-app-version",
        "1.0",
        "--uid",
        "1111"
     ]
     ....
containers:
   - name: main-app-container
     image: main-app:1.0
     securityContext:
        runAsUser: 1111
        runAsGroup: 1111
     ...
```
Then when the `main-app` gets updated to `2.0` , the spec looks like the following: 
```
initContainers:
   - name: pre-app-container-setup
     image: pre-app-container-setup:100.0
     command: ["utils-bin"]
     args: [
        "--main-app-version",
        "2.0",  # changed 
        "--uid",
        "1111"
     ]
     ....
containers:
   - name: app-container
     image: main-app:2.0
     securityContext:
        runAsUser: 1111
        runAsGroup: 1111
     ...
```
With this spec, even though we can _supposably_  do in-place update for the main app container, but due to the change in the init container `pre-app-container-setup`, it causes the entire pod to be bounced, which defeats the purpose of in-place benefit of the app container. 

What could have happened could be, init container gets in place update (restarted)  followed by the main app container gets restarted, rather than the pod gets Recreated.

@Fei-Guo ^^ WDYT?

**Comment by Fei-Guo** [comment]
Inplace updating init container may work (kubelet seems to support it). One problem is that init container restart cannot trigger main containers to restart automatically. The real benefit of supporting this is not very clear as you may need to update initcontainer image and main container images at the same time, which should not be a common practice. 

The code change to support this feature is nontrivial. If you are willing to give a try, it is certainly welcome.

**Comment by wangyanghack** [comment]
@wgitg hi, i know about your mean. But just image changes for container can in-place update in cloneset. So are you sure to want to implement init container in-place update for command changes? Or you want to implement init container in-place update for image changes?

**Comment by FillZpp** [comment]
@wgitg Hi, we are still considering this feature. Do you think it is reasonable that `initContainers` can be updated in-place with no `containers` restarted?

**Comment by wgitg** [comment]
@FillZpp Thanks for reconsidering this feature! As for restarting the `containers` or not, it may depend. But the bottomline is
1. pod is still on that node. 
2. `containers` can execute on the latest data/updates that init containers are set up to do.

That said, to do 2, does it make sense we either restart all the containers or we can add some label that only the matching container will be restarted once the the init container that have the same labels?

**Comment by FillZpp** [comment]
> or we can add some label that only the matching container will be restarted once the the init container that have the same labels

About this, you can use the feature [in-place udpate env from metadata](https://openkruise.io/blog/openkruise-1.0/#1-inplace-update-for-environments) released in v1.0.

However, when you trigger an init container and a container to be updated, I think we have no way to control their sequence, which means the container might be restarted before the init container has been updated. That's the problem.

**Comment by wgitg** [comment]
@FillZpp  Awesome! We can potentially use the env metadata and annotation per environment variable (as used before). Two more follow up questions: 
1. Does this feature work for both init container and regular containers? 

Specifically, if either or both the init containers and the regular containers use the environment meta and the value is changed in the annotation, are all the following behavior true?
a. the pod will NOT be recreated
b. the init containers/application containers are just restarted, no IP changes
c. the execution ordering of init container and app containers are guaranteed

2. The feature is only available in v1.0.0, which has a bunch of other bumps, such as k8s api bump from 1.18 to 1.20, controller-runtime bump, CRD API bump... The thing is when we upgraded openkruise two months ago, the v1.0.0 is still in alpha, and our k8s cluster api is yet to upgraded to 1.18 from 1.17 and so we upgraded kruise to v0.9.0 so far. So if it is not too much ask, is it possible to make the feature available in v0.9.0, or we must bump everything to use v1.0.0?

**Comment by wgitg** [comment]
@jian-he @FillZpp  @Fei-Guo  can any one of you share some insights on the above comments/questions?

**Comment by FillZpp** [comment]
> can any one of you share some insights on the above comments/questions?

@wgitg Sorry for this late reply. May I ask could you speak Chinese? If so, we can talk about this proposal on [the community meeting this week](https://shimo.im/docs/gXqmeQOYBehZ4vqo#anchor-ltkM). WDYT?

**Comment by wgitg** [comment]
@FillZpp Yes. I can speak Chinese. And I'd love to talk about the proposal more.. but I am located in the US bay area in PST time,  is it possible to to make the meeting more US time friendly? :)

**Comment by FillZpp** [comment]
Oh, I'm sorry currently we can not change its time for some of our community members in China can only attend in the evening. So I'm going to answer your questions above, and we can also have a discussion using zoom meeting with less people if you need.

> Does this feature work for both init container and regular containers?

No, same as image in-place update, the env in-place update with Downward API can only work for regular containers.

> Specifically, if either or both the init containers and the regular containers use the environment meta and the value is changed in the annotation, are all the following behavior true?

When you update regular containers in-place, the pod will NOT be recreated and the containers are just restarted, no IP changes, etc. But if you trigger multiple regular containers to in-place update at once, the execution ordering are NOT guaranteed. They may be updated together.

> The feature is only available in v1.0.0, which has a bunch of other bumps, such as k8s api bump from 1.18 to 1.20, controller-runtime bump, CRD API bump...

The truth is it only updates the dependencies of our code and does not mean it can only be used in 1.20 cluster. Actually, Kruise 1.0 can be deployed in clusters with k8s version >= 1.16. Open any one of [Kruise Pull Requests](https://github.com/openkruise/kruise/pulls)  and you will see our e2e cases should run in 1.22 and 1.16 clusters. They must all be successful before the PR being merged.

**Comment by wgitg** [comment]
@FillZpp  Got it. Thanks for the information on the kruise 1.0 on cluster version >= 1.16. 
And we will update the plan on leveraging the plan in the team due to the in-place update with the environment metadata only works for regular container but not the init containers. :(
yeah, we can follow up more on Zoom...

**Comment by wgitg** [comment]
Thanks for the engaging discussion and meetup,  @FillZpp and others. :+1:

Given the complexity and complication to support in-place update for init containers, the [alternative proposal](https://github.com/openkruise/kruise/issues/901) can help us to truly achieve the in-place update. We may close this issue.

**Comment by FillZpp** [comment]
closed by #901
